src/include/common/perf_monitor.h

```diff
@@ -0,0 +1,203 @@
+#pragma once
+
+#if __APPLE__
+// nothing to include since it doesn't support perf events
+#else
+#include <linux/perf_event.h>
+#include <sys/ioctl.h>
+#include <sys/syscall.h>
+#endif
+#include <unistd.h>
+#include <array>
+#include <cstring>
+#include <iostream>
+#include "common/macros.h"
+
+namespace terrier::common {
+/**
+ * Wrapper around hw perf events provided by the Linux kernel. Instantiating and destroying PerfMonitors are a bit
+ * expensive because they open multiple file descriptors (read: syscalls). Ideally you want to keep a PerfMonitor object
+ * around for a portion of code you want to profile, and then just rely on Start() and Stop().
+ */
+class PerfMonitor {
+ public:
+  /**
+   * Represents the struct read_format with PERF_FORMAT_GROUP enabled, PERF_FORMAT_TOTAL_TIME_ENABLED and
+   * PERF_FORMAT_TOTAL_TIME_RUNNING disabled. http://www.man7.org/linux/man-pages/man2/perf_event_open.2.html
+   */
+  struct PerfCounters {
+    /**
+     * Number of events read. This should be PerfMonitor::NUM_HW_EVENTS if working, 0 otherwise
+     */
+    uint64_t num_events_;
+    /**
+     * Total cycles. Be wary of what happens during CPU frequency scaling.
```

Comment:
What does this mean? Can you be a little bit more specific?

Reply:
Just copying from the man page.